### PR TITLE
CountInLoop Rule initial implementation

### PR DESCRIPTION
**Pull request by eeree**
Rule applies whenever in loop (for/while/do) we use count() or sizeof() functions.
```
for ($i = 0; count($_GET); $i++) {
  // ...
}
```

### CODE REVIEW

**Comment by ravage84** [comment]
If you rebase your branch, it will contain https://github.com/phpmd/phpmd/commit/649774d55f16499cf8a2cdaaa6d739dfabc181c2 whcih should silence Stickler-CI.

**Comment by eeree** [comment]
@ravage84 
No more WIP.

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+use PHPMD\AbstractRule;
+use PHPMD\Node\ASTNode;
+use PHPMD\Rule\ClassAware;
+
+/**
+ * Count In Loop Expression Rule
+ *
+ * Performs a scan to check if loops use
+ * count() or sizeof() in expressions.
+ * Works with:
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: White line. Then "Checks for"

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+use PDepend\Source\AST\AbstractASTNode;
+use PHPMD\AbstractNode;
+use PHPMD\AbstractRule;
+use PHPMD\Node\ASTNode;
+use PHPMD\Rule\ClassAware;
+
+/**
+ * Count In Loop Expression Rule
+ *
+ * Performs a scan to check if loops use
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Rewrite

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+ * @author Kamil Szymanski <kamilszymanski@gmail.com>
+ */
+class CountInLoopExpression extends AbstractRule implements ClassAware
+{
+    /**
+     * List of functions to search against
+     *
+     * @var array
+     */
+    private $violatingFunctions = array('count', 'sizeof');
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: $unwantedFunctions as we have that already somewhere else?

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    /**
+     * Functions in classes tends to be name-spaced
+     *
+     * @var string
+     */
+    private $namespaceName = '';
+
+    /**
+     * Gets a list of loops in node and iterates over them
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: In a node

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+     * @var array
+     */
+    private $violatingFunctions = array('count', 'sizeof');
+
+    /**
+     * List of already processed functions
+     *
+     * @var array
+     */
+    private $processedFunctions = array();
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Protected

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+    /**
+     * Functions in classes tends to be name-spaced
+     *
+     * @var string
+     */
+    private $namespaceName = '';
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Protected

Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: $currentNamespace as it tells more what it is?

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        }
+    }
+
+    /**
+     * Scans for expressions and count() or sizeof() functions inside.
+     * If found, triggers violation
+     *
+     * @param AbstractNode $loop Loop statement to look against
+     */
+    private function findPossibleViolations(AbstractNode $loop)
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Protected

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    /**
+     * Checks if expression node in a direct child of loop
+     *
+     * @param AbstractNode $loop
+     * @param ASTNode $expression
+     * @return bool
+     */
+    private function isDirectChild(AbstractNode $loop, ASTNode $expression)
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Protected

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        return $this->getHash($expression->getParent()->getNode()) !== $this->getHash($loop->getNode());
+    }
+
+    /**
+     * Generates unique hash for given node
+     *
+     * @param AbstractASTNode $node
+     * @return string
+     */
+    private function getHash(AbstractASTNode $node)
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Protected

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        );
+    }
+
+    /**
+     * Checks if given function exists in violations array
+     *
+     * @param ASTNode $function
+     * @return bool
+     */
+    private function isViolatingRule(ASTNode $function)
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Protected

Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: isUnwantedFunction?

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+            $node->getStartLine(),
+            $node->getEndLine(),
+            $node->getStartColumn(),
+            $node->getEndColumn(),
+            $node->getImage()
+        );
+    }
+
+    /**
+     * Checks if given function exists in violations array
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Checks the given function against the list of unwanted functions

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    private function isDirectChild(AbstractNode $loop, ASTNode $expression)
+    {
+        return $this->getHash($expression->getParent()->getNode()) !== $this->getHash($loop->getNode());
+    }
+
+    /**
+     * Generates unique hash for given node
+     *
+     * @param AbstractASTNode $node
+     * @return string
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Can you add an example of such a hash as Long desc?

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+     * @param ASTNode $expression
+     * @return bool
+     */
+    private function isDirectChild(AbstractNode $loop, ASTNode $expression)
+    {
+        return $this->getHash($expression->getParent()->getNode()) !== $this->getHash($loop->getNode());
+    }
+
+    /**
+     * Generates unique hash for given node
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: a hash
a node

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+                }
+
+                $this->addViolation($loop, array($function->getImage(), $loop->getImage()));
+                $this->processedFunctions[$hash] = true;
+            }
+        }
+    }
+
+    /**
+     * Checks if expression node in a direct child of loop
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Checks whether... ?

File: src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php (hunk 7 above)
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/CountInLoopExpression.php]: Why just "possible"?

**Pull request merged**